Add back in derived variables, reorder dropdown

nicolejkeeney opened the pull request:
In the original catalog restructure PR, I realized I forgot to add in the functionality to derive the hourly variables: precipitation total, wind magnitude, and relative humidity. I add that back in here, branched off of Eric's branch `code_cleanup` (which in turn is branched off of `catalog_restructure`.) This fix required me to change up data_loaders.py a bit. 

I also reordered the variables in variable_catalog.csv and cleaned up the descriptions for hourly and daily/monthly qv, along with snowfall. 

**To test:** 
Try selecting and loading the derived hourly variables and make sure the data looks correct, and the unit conversion options for those variables look good. Confirm that you agree with the ordering for the daily/monthly variables.

elehmer commented:
Ahh, this would explain why I was getting a kernel crash on those variables...

elehmer commented:
Can you merge the `code_cleanup` branch with this one as I've done a 2nd pass today so this is out of date.

nicolejkeeney commented:
> Can you merge the `code_cleanup` branch with this one as I've done a 2nd pass today so this is out of date.

I believe I was able to successfully merge them! See the commit above.

vicford commented on climakitae/data/variable_catalog.csv:
```diff
@@ -1,49 +1,52 @@
-t2,2-meter temperature ,daily/monthly,K,Air Temperature at 2m,Temperature of the air 2m above Earth's surface. This is the measure of air temperature used for most modeling applications.,ae_orange,TRUE
+sw_sfc,shortwave flux at the surface,daily/monthly,W/m2,Shortwave flux at the surface,Shortwave radiation from the sun that reaches Earth's surface. ,ae_orange,TRUE
+lw_sfc,longwave flux at the surface,daily/monthly,W/m2,Longwave flux at the surface,Longwave radiation emitted by Earth's surface.,ae_orange,TRUE
+sh_sfc,sensible heat flux at the surface,daily/monthly,W/m2,Sensible heat flux at the surface,"The movement of energy generated by a difference in temperatures between the surface and the atmosphere. By convention, positive sensible heat flux is directed from the surface up into the atmosphere. ",ae_diverging,TRUE
```
I'd recommend sensible heat flux, latent heat flux, and ground heat flux be grouped together, in that order in the dropdown list for the daily data as they're similar data variables and are often analyzed together.

vicford replied:
![Screen Shot 2022-10-26 at 12 16 36 PM](https://user-images.githubusercontent.com/106171203/198079793-0eb4a595-02df-407a-8ba8-d4ab912e2650.png)

nicolejkeeney replied:
> I'd recommend sensible heat flux, latent heat flux, and ground heat flux be grouped together, in that order in the dropdown list for the daily data as they're similar data variables and are often analyzed together.

Fixed!

vicford commented on climakitae/data/variable_catalog.csv:
```diff
@@ -1,49 +1,52 @@
+sw_sfc,shortwave flux at the surface,daily/monthly,W/m2,Shortwave flux at the surface,Shortwave radiation from the sun that reaches Earth's surface. ,ae_orange,TRUE
+lw_sfc,longwave flux at the surface,daily/monthly,W/m2,Longwave flux at the surface,Longwave radiation emitted by Earth's surface.,ae_orange,TRUE
+sh_sfc,sensible heat flux at the surface,daily/monthly,W/m2,Sensible heat flux at the surface,"The movement of energy generated by a difference in temperatures between the surface and the atmosphere. By convention, positive sensible heat flux is directed from the surface up into the atmosphere. ",ae_diverging,TRUE
+prec_snow,snow precipitation,daily/monthly,mm,Snowfall,Portion of the total precipitation in the form of snow. ,ae_blue,TRUE
```
Confirming that for monthly/daily snowfall that it is just the snow portion with no ice; whereas for hourly it is snowfall with both snow and ice?

nicolejkeeney replied:
Will follow up again with Brian

vicford commented:
Hourly RH units are acting not quite right:
- The 0 to 100 scale is being represented as 0 to 1 (picture 1)
![Screen Shot 2022-10-26 at 12 05 04 PM](https://user-images.githubusercontent.com/106171203/198078325-3b41da85-7f67-4b51-908b-328331b410db.png)


- The fractional scale is displaying water vapor mixing ratio instead (picture 2)
![Screen Shot 2022-10-26 at 12 07 47 PM](https://user-images.githubusercontent.com/106171203/198078343-1ef0e69c-4808-4653-9549-c9facb39e163.png)

vicford commented on climakitae/data/variable_catalog.csv:
```diff
@@ -1,49 +1,52 @@
+t2,temp at 2 m,hourly,K,Air Temperature at 2m,Temperature of the air 2m above Earth's surface. This is the measure of air temperature used for most modeling applications.,ae_orange,TRUE
+precip_tot_derived,Precipitation (total),hourly,mm,Precipitation (total),Total precipitation. Computed by summing total gridscale precipitation and total cumulus precipitation.,ae_blue,TRUE
+rh_derived,Relative Humidity,hourly,[0 to 100],Relative Humidity,The percentage of water vapor (i.e. moisture) relative to the amount of water vapor needed to saturate the air at the current air temperature and pressure.,ae_blue,TRUE
+wind_mag_derived,Wind Magnitude,hourly,m s-1,Wind Magnitude at 10m,Wind speed at 10 meters above the Earth's surface. Computed by taking the vector magnitude of the west-east (u) component of wind and the north-south (v) component of wind.,ae_orange,TRUE
```
For wind magnitude, which we derive, everything looks fine when the data is called in at hourly timescales except for the variable_id which is set to u10 (which we define as for the west-east component of the wind). 

This is fine and somewhat confusing difference but "u10" is technically okay as long as it's a capital U, so "U10" for wind magnitude. But lower-case "u10" is specifically for E-W wind. But since I can see "wind_mag_derived" that would be better to include so things don't get confusing.

vicford replied:
To clarify: my preference for the wind magnitude variable is to have the variable_id be set to "wind_mag_derived"

elehmer commented:
> Hourly RH units are acting not quite right:
> 
> * The 0 to 100 scale is being represented as 0 to 1 (picture 1)
>   ![Screen Shot 2022-10-26 at 12 05 04 PM](https://user-images.githubusercontent.com/106171203/198078325-3b41da85-7f67-4b51-908b-328331b410db.png)
> * The fractional scale is displaying water vapor mixing ratio instead (picture 2)
>   ![Screen Shot 2022-10-26 at 12 07 47 PM](https://user-images.githubusercontent.com/106171203/198078343-1ef0e69c-4808-4653-9549-c9facb39e163.png)

Line 8 of `variable_catalog.csv` has an unquoted `'`. That might be throwing off the lists...

vicford reviewed (changes requested):
Total precipitation looks fine to me, but please see notes on RH and wind

elehmer commented:
> Hourly RH units are acting not quite right:
> 
> * The 0 to 100 scale is being represented as 0 to 1 (picture 1)
>   ![Screen Shot 2022-10-26 at 12 05 04 PM](https://user-images.githubusercontent.com/106171203/198078325-3b41da85-7f67-4b51-908b-328331b410db.png)
> * The fractional scale is displaying water vapor mixing ratio instead (picture 2)
>   ![Screen Shot 2022-10-26 at 12 07 47 PM](https://user-images.githubusercontent.com/106171203/198078343-1ef0e69c-4808-4653-9549-c9facb39e163.png)

I'm not seeing this behavior. This is what I'm getting with (hourly, 1980-2015, Historical Climate, Relative Humidity, [0-100] units, 45km, CA Counties-Alameda) :
![image](https://user-images.githubusercontent.com/9967770/198138780-97518f8e-427c-4ac5-9c45-8545f339c5aa.png)

elehmer commented:
> > Hourly RH units are acting not quite right:
> > 
> > * The 0 to 100 scale is being represented as 0 to 1 (picture 1)
> >   ![Screen Shot 2022-10-26 at 12 05 04 PM](https://user-images.githubusercontent.com/106171203/198078325-3b41da85-7f67-4b51-908b-328331b410db.png)
> > * The fractional scale is displaying water vapor mixing ratio instead (picture 2)
> >   ![Screen Shot 2022-10-26 at 12 07 47 PM](https://user-images.githubusercontent.com/106171203/198078343-1ef0e69c-4808-4653-9549-c9facb39e163.png)
> 
> I'm not seeing this behavior. This is what I'm getting with (hourly, 1980-2015, Historical Climate, Relative Humidity, [0-100] units, 45km, CA Counties-Alameda) : ![image](https://user-images.githubusercontent.com/9967770/198138780-97518f8e-427c-4ac5-9c45-8545f339c5aa.png)

Ahh, ok I was able to replicate when taking a time slice of the data. Let's look into that portion of the code...

elehmer commented:
> > > Hourly RH units are acting not quite right:
> > > 
> > > * The 0 to 100 scale is being represented as 0 to 1 (picture 1)
> > >   ![Screen Shot 2022-10-26 at 12 05 04 PM](https://user-images.githubusercontent.com/106171203/198078325-3b41da85-7f67-4b51-908b-328331b410db.png)
> > > * The fractional scale is displaying water vapor mixing ratio instead (picture 2)
> > >   ![Screen Shot 2022-10-26 at 12 07 47 PM](https://user-images.githubusercontent.com/106171203/198078343-1ef0e69c-4808-4653-9549-c9facb39e163.png)
> > 
> > 
> > I'm not seeing this behavior. This is what I'm getting with (hourly, 1980-2015, Historical Climate, Relative Humidity, [0-100] units, 45km, CA Counties-Alameda) : ![image](https://user-images.githubusercontent.com/9967770/198138780-97518f8e-427c-4ac5-9c45-8545f339c5aa.png)
> 
> Ahh, ok I was able to replicate when taking a time slice of the data. Let's look into that portion of the code...

Actually, it is when you back up to `app.select()` and run it again. Even if you don't change anything. This is some sort of panel issue I do believe...

vicford commented:
> Actually, it is when you back up to `app.select()` and run it again. Even if you don't change anything. This is some sort of panel or settings dictionary issue I do believe...

That's curious to me... I didn't re-select app.select() twice in my testing... but just a single pass through. I did modify some selections in the pre-existing panel, but I didn't re-run the app.select() cell.... hmmmm

elehmer commented:
> > Actually, it is when you back up to `app.select()` and run it again. Even if you don't change anything. This is some sort of panel or settings dictionary issue I do believe...
> 
> That's curious to me... I didn't re-select app.select() twice in my testing... but just a single pass through. I did modify some selections in the pre-existing panel, but I didn't re-run the app.select() cell.... hmmmm

Yes, I think we are talking about the same thing. Start from fresh kernel, go down to `app.select` pick relevant variables, get data, then pick new items from panel or just run from `app.select` cell onwards again and SPLAT!

vicford commented:
**In some additional testing this morning on this perplexing RH error:** 

For 0-100 scale
- Monthly is okay
- Daily is okay
- Hourly does not display on 0-100 scale (it looks as if the fractional conversion is taking place)

For fractional scale
- Monthly RH returns Q2 at g/kg
- Daily RH returns Q2 at g/kg
- Hourly RH returns Q2 at kg/kg

So there might be multiple issues taking place at the hourly scale with data retrieval + unit conversions. 

**For total precipitation:**
- Monthly returns both mm and inches with no problem
- Daily returns mm/day regardless of unit selection, i.e., does not convert to inches
- Hourly _sometimes_ does not retrieve inches (see below error), but when it does it does correctly convert to inches

For hourly/daily, after several data selections of working in select, I've been getting the following error for the hour selection, even though there is a large enough time slice and location selection (1996-1998 + CA). Restarting the kernel "fixes" this, but this doesn't address what might be going on behind the scenes?
```
/srv/conda/envs/notebook/lib/python3.9/site-packages/intake_esm/search.py:106: UserWarning: Query returned zero results.
  warn(message)
/srv/conda/envs/notebook/lib/python3.9/site-packages/intake_esm/core.py:882: UserWarning: There are no datasets to load! Returning an empty dictionary.
  warn('There are no datasets to load! Returning an empty dictionary.')
ValueError: Length of values (0) does not match length of index (9)
```

**For wind magnitude**:
- Only hourly now provides "wind magnitude at 10m" is that correct? Why are we no longer providing this for monthly/daily?
- For hourly in **m/s** the returned variable_id is "u10" (east-west wind) ~
- For hourly in **knots** the returned variable_id is "v10" (north-south wind) ~
- The unit conversions themselves do return the correct values for a converted unit

~ I checked both, and it's not returning E-W or N-S wind, just the variable_id is wrong

nicolejkeeney commented:
> **For total precipitation:**
> 
> * Monthly returns both mm and inches with no problem
> * Daily returns mm/day regardless of unit selection, i.e., does not convert to inches
> * Hourly _sometimes_ does not retrieve inches (see below error), but when it does it does correctly convert to inches

I believe I have resolved the issue with daily conversions!

nicolejkeeney commented:
> **For wind magnitude**:
> 
> * Only hourly now provides "wind magnitude at 10m" is that correct? Why are we no longer providing this for monthly/daily?
> * For hourly in **m/s** the returned variable_id is "u10" (east-west wind) ~
> * For hourly in **knots** the returned variable_id is "v10" (north-south wind) ~
> * The unit conversions themselves do return the correct values for a converted unit
> 
> ~ I checked both, and it's not returning E-W or N-S wind, just the variable_id is wrong

I believe I have just fixed this issue with my most recent commits! :D

nicolejkeeney commented:
Hi @elehmer and @vicford, I accidentally fixed all the issues in a separate branch that is branched off of this branch... oops! Please test that branch when you can, then we can merge that branch into this one :D 

The other branch: https://github.com/cal-adapt/climakitae/pull/115

vicford commented:
> Please test that branch when you can, then we can merge that branch into this one

Almost there!
**RH**:
Hourly: 0-100 scale, fixed!
Hourly: fractional scale -- still on a 0-100 scale instead
Daily: fractional scale, fixed!
Monthly: fractional scale, fixed!

**Total Precip**:
Hourly error, didn't get at all!
Daily for both mm and inches -- the conversion is correct, but the values are super small... like less than 1, whereas for monthly and hourly data for mm were in the 100-800s, and ~25 inches

**Wind**:
Variable id for hourly wind mag, fixed!
No monthy/daily

nicolejkeeney commented:
> **Total Precip**: Hourly error, didn't get at all! Daily for both mm and inches -- the conversion is correct, but the values are super small... like less than 1, whereas for monthly and hourly data for mm were in the 100-800s, and ~25 inches

There's actually no conversion happening for the daily/monthly data. Whatever data you see is straight from the catalog (unless its been converted to inches). I.e. we don't compute total precip for daily/monthly, we already have that data.

nicolejkeeney commented:
> Almost there! **RH**: Hourly: 0-100 scale, fixed! Hourly: fractional scale -- still on a 0-100 scale instead Daily: fractional scale, fixed! Monthly: fractional scale, fixed!

I believe I fixed the hourly conversion just now :D

nicolejkeeney commented:
> Hi @elehmer and @vicford, I accidentally fixed all the issues in a separate branch that is branched off of this branch... oops! Please test that branch when you can, then we can merge that branch into this one :D
> 
> The other branch: #115

I merged #115 with this branch a minute ago so all the changes are now in this branch :D

vicford reviewed (approved):
Looks good to me Nicole. Thanks for all of these detailed changes!

bethmints commented:
General variable notes:
- Seeing negative values for hourly wind magnitude in knots. `variable_id` says v10, though the other attributes match expectations for wind magnitude. 
- (niggling issue) it's called "wind magnitude" for hourly and "wind speed" for daily/monthly.
- It's been brought up, but we do see some barely supersaturated values in hourly RH. 
- For precipitation variables, explicitly mention that it's accumulated over the water year. It took me a hot minute to interpret the app.view() graph, lol
- For daily and monthly sfc pressure in inHg, I'm getting values around 3000. Perhaps related to hPa -> Pa conversion?

Personal preferences on catalog order:
- Group all precipitation values together (total, cumulus, grid-scale, snow)
- Group all temperature variables together (2m, skin, max/min)
- Group RH and qv together (and have those after temperature variables)

nicolejkeeney commented:
> * For daily and monthly sfc pressure, all values are two orders of magnitude too high (inHg _O_(1e3), hPa/mb _O_(1e5), Pa _O_(1e7)); perhaps mixed up with hPa -> Pa conversion?

Oh my gosh, great catch. Thanks, Beth. I had accidentally multiplied by 100 instead of divided by 100 when going from hPa --> Pa. It's fixed now!

nicolejkeeney commented:
> * Seeing negative values for hourly wind magnitude in knots. `variable_id` says v10, though the other attributes match expectations for wind magnitude.
> * ~(niggling issue) it's called "wind magnitude" for hourly and "wind speed" for daily/monthly.~ okay, it's max speed for daily/monthly. still maybe should be "wind speed" for hourly?

I just fixed these both!! Great catch, Beth!!

nicolejkeeney commented:
> * It's been brought up, but we do see some supersaturated values in hourly RH.

Not sure how to fix this. Could be a number of reasons. Do you think we can merge this PR with this potential issue?

bethmints commented:
> > * It's been brought up, but we do see some supersaturated values in hourly RH.
> 
> Not sure how to fix this. Could be a number of reasons. Do you think we can merge this PR with this potential issue?

from what I've seen, it's not too unusual for RH > 100 to occasionally pop up in model output. corrections usually involve setting these values to 100. we should bring this up with the larger group, however.

nicolejkeeney commented:
> > * For daily and monthly sfc pressure, all values are two orders of magnitude too high (inHg _O_(1e3), hPa/mb _O_(1e5), Pa _O_(1e7)); perhaps mixed up with hPa -> Pa conversion?
> 
> Oh my gosh, great catch. Thanks, Beth. I had accidentally multiplied by 100 instead of divided by 100 when going from hPa --> Pa. It's fixed now!

Update: The hPa --> Pa conversion is messed up due to an issue with the catalog data having an incorrectly assigned native variable. Waiting on that fix to merge this PR.
